axelor-base/src/main/java/com/axelor/apps/base/service/ImportExportTranslationServiceImpl.java

```diff
@@ -0,0 +1,357 @@
+                    + "on base_import_export_translation_language_set.base_import_export_translation = base_import_export_translation.id "
+                    + "left join base_language "
+                    + "on base_import_export_translation_language_set.language_set = base_language.id "
+                    + "where base_import_export_translation.id = :curId ")
+            .setParameter("curId", importExportTranslationId);
+    return (List<String>) languageQuery.getResultList();
+  }
+
+  /**
+   * This method imports the translation file and saves each record into ImportExportTranslation
+   * obj.
+   *
+   * @param importExportTranslation
+   * @return
+   */
+  @Transactional
+  @Override
+  public Path importTranslations(ImportExportTranslation importExportTranslation) {
+    errorNumber = 0;
+    Path filePath = MetaFiles.getPath(importExportTranslation.getUploadFile());
+    char separator = ';';
+    List<String[]> data = null;
+    try {
+      data = CsvHelper.cSVFileReader(filePath.toString(), separator);
+    } catch (Exception e) {
+      Logger logger = LoggerFactory.getLogger(getClass());
+      logger.error("Read CSV file failed.", e);
+    }
+    if (data == null) return null;
+    List<String> headers = Arrays.asList(data.get(0));
+    int dataLineNumber = 0;
+    for (int i = 1; i < data.size(); i++) {
+      List<String> dataLine = Arrays.asList(data.get(i));
+      try {
+        dataLineNumber += insertOrUpdateTranslation(dataLine, headers);
+      } catch (Exception e) {
+        Logger logger = LoggerFactory.getLogger(getClass());
+        logger.error("Error processing data line: " + dataLine, e);
+        errorNumber++;
+      }
+    }
+    ImportExportTranslationHistory importExportTranslationHistory =
+        new ImportExportTranslationHistory();
+    importExportTranslation =
+        importExportTranslationRepository.find(importExportTranslation.getId());
+    importExportTranslationHistory.setActionType("Import");
+    importExportTranslationHistory.setRecordNumber(dataLineNumber);
+    importExportTranslationHistory.setErrorNumber(errorNumber);
+    importExportTranslationHistory.setImportExportTranslation(importExportTranslation);
+    importExportTranslation.addImportExportTranslationHistoryListItem(
+        importExportTranslationHistory);
+    importExportTranslationRepository.save(importExportTranslation);
+    return filePath;
+  }
+
+  /**
+   * This method insert/update each dataLine.
+   *
+   * @param dataLine
+   * @param headers
+   * @return total records imported or updated for the dataLine.
+   */
+  @Transactional
+  protected int insertOrUpdateTranslation(List<String> dataLine, List<String> headers)
+      throws Exception {
+    int importNumber = 0;
+    String key = dataLine.get(0);
+
+    if (dataLine.size() != headers.size()) {
+      throw new Exception("CSV file line has different number of columns against headers");
+    }
+
+    for (int i = 1; i < dataLine.size(); i++) {
```

Comment:
use `for (String line : dataLine)`

Reply:
Because we use the index to get particular headers, we should not replace it with a for-each loop.